core/src/preferences/PreferencesBackend.cpp

```diff
@@ -196,8 +223,96 @@ namespace ledger {
             _cipher = Option<AESCipher>::NONE;
         }
 
+        void PreferencesBackend::resetEncryption(
+            const std::shared_ptr<api::RandomNumberGenerator>& rng,
+            const std::string& newPassword,
+            const std::string& oldPassword
+        ) {
+            // turn on encryption with the old password first
+            setEncryption(rng, oldPassword);
+
+            // from now on, reading data will use the old password, but we want to persist with a
+            // brand new cipher; create it here
+            auto newSalt = createNewSalt(rng);
+            auto newCipher = Option<AESCipher>(AESCipher(rng, newPassword, newSalt, PBKDF2_ITERS));
+
+            // we will also need a brand new leveldb to implement a somewhat atomic database swap;
+            // the idea is to read from the old leveldb and write to this new one; when we’re done,
+            // we just “swap” the database and remove the old one; this swap is needed so that we
+            // don’t corrupt the old database if we crash or get interrupted while in the middle of
+            // the process of resetting
+            auto tempDBName = _dbName + "__temp_copy__";
```

Comment:
I think we can do this without additional DB, just push all changes to one batch https://github.com/google/leveldb/blob/master/doc/index.md#atomic-updates , both delete old values and put new then make one Write(batch) that will do the things, this way we push the burden to atomically swap to levelDB.

Reply:
Nice idea. I’ll change it to reflect your idea. It’ll be way more lightweight.

Reply:
Done. That was nice to rewrite the code with this feature. 😄